app/models/concerns/manage_user_roles.rb

```diff
@@ -0,0 +1,44 @@
+module ManageUserRoles
+  extend ActiveSupport::Concern
+
+  included do
+
+    def update_roles_for_user(user, role_ids)
+      return unless self.can_update_roles_for_user?(user)
+
+      allowed_role_ids = self.update_allowed_on_roles.where(id: role_ids).pluck(:id)
+      user.create_or_remove_user_roles(allowed_role_ids)
+    end
+
+    def can_update_roles_for_user?(other_user)
+      self != other_user &&
+      self.max_role_level >= other_user.max_role_level
+    end
+
+    def max_role_level
+      self.roles.maximum("level") || 0
+    end
+
+    def update_allowed_on_roles
```

Comment:
```suggestion
    # allow the user to only update roles up to and including their level
    def allowed_roles_to_update
```

Reply:
Even better, this could be a scope on Role

`Role.max_role_level(10)` provides all roles up to and including level 10 roles

`scope :max_role_level, ->(level) { where('level <?', level) }`